src/System.Private.CoreLib/shared/System/Type.Helpers.cs

```diff
@@ -129,27 +129,27 @@ public virtual Type[] FindInterfaces(TypeFilter filter, object filterCriteria)
                     cnt++;
             }
             if (cnt == c.Length)
-                return c;
+                return c!;
```

Comment:
Why is the exclamation point necessary here?  Surely the compiler can see that we just dereferenced `c` and thus it can't be `null`.  Was this actually necessary, or is it a compiler analysis bug?
cc: @jaredpar (Jared, what's the best way to ask these questions... continue tagging you?  Tagging you and others on your team?  Just tagging others on your team?  etc.)

Reply:
Because of assigning null to `c[i] = null;` on row 127 we need to declare c[] as `Type?[] c` which leads compiler error in this row `return c;` without !
```csharp
            Type?[] c = GetInterfaces();
            int cnt = 0;
            for (int i = 0; i < c.Length; i++)
            {
                if (!filter(c[i], filterCriteria))
                    c[i] = null;
                else
                    cnt++;
            }
            if (cnt == c.Length)
                return c!;
```

Reply:
The code is doing this:
```C#
if (cnt == c.Length) return c;
```
There is no way that `return c` could happen if `c` is null, because we just dereferenced it to access `c.Length`.  The compiler should absolutely be able to see that.

Reply:
This doesn't make sense, c is not nullable, what is nullable are the contents of c here. I don't think the ! should be needed here.

Reply:
exactly, type `Type[]` is not matching with `Type?[]`

Reply:
I see.  The `!` isn't casting away the nullness of `c`, it's casting away the nullness of the elements `c` contains.  That's... a bit odd.

Reply:
Agree, that is confusing. Are you 100% this is needed or is it just something you added accidentally? If so what is the error you're getting? So that we can get the compiler team look at it.

Reply:
The `!` operator does two things: it casts away top-level nullability, and it suppresses any warnings. So in this case, the result of `c!` is still `Type?[]` but the nullability mismatch warning is suppressed.

---
In reply to: [270495133](https://github.com/dotnet/coreclr/pull/23489#discussion_r270495133) [](ancestors = 270495133)

Reply:
@safern  this is the error: 

![image](https://user-images.githubusercontent.com/45579687/55250209-e5cbcc80-520a-11e9-863b-d72a4ac6ad79.png)

Reply:
Yeah, just what @cston described. So then we do need to use ! and live with it.